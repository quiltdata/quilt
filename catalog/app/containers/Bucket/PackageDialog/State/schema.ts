import type { ErrorObject } from 'ajv'
import * as React from 'react'
import * as R from 'ramda'

import * as AWS from 'utils/AWS'
import {
  JsonSchema,
  makeSchemaDefaultsSetter,
  makeSchemaValidator,
} from 'utils/JSONSchema'
import * as Request from 'utils/useRequest'
import * as Types from 'utils/types'
import * as workflows from 'utils/workflows'

import { metadataSchema, objectSchema } from '../../requests'

export type SchemaStatus =
  | { _tag: 'idle' }
  | { _tag: 'loading' }
  | { _tag: 'error'; error: Error }
  | { _tag: 'ready'; schema?: JsonSchema }

export const Idle = { _tag: 'idle' as const }
export const Loading = { _tag: 'loading' as const }
export const Err = (error: Error) => ({ _tag: 'error' as const, error })
export const Ready = (schema?: JsonSchema) => ({ _tag: 'ready' as const, schema })

export function mkMetaValidator(schema?: JsonSchema) {
  const schemaValidator = makeSchemaValidator(schema)
  return function validateMeta(value: Types.Json): (ErrorObject | Error)[] | undefined {
    const jsonObjectErr = value && !R.is(Object, value)
    if (jsonObjectErr) {
      return [new Error('Metadata must be a valid JSON object')]
    }

    const setDefaults = makeSchemaDefaultsSetter(schema)
    const errors = schemaValidator(setDefaults(value))
    if (errors.length) return errors
  }
}

export function useMetadataSchema(workflow?: workflows.Workflow): SchemaStatus {
  const s3 = AWS.S3.use()
  const schemaUrl = workflow?.schema?.url
  const req = React.useCallback(() => metadataSchema({ s3, schemaUrl }), [schemaUrl, s3])
  const result = Request.use(req, !!schemaUrl)

  if (!schemaUrl) return Ready()

  if (result === Request.Idle) return Idle
  if (result === Request.Loading) return Loading
  if (result instanceof Error) return Err(result)

  return Ready(result)
}

export function useEntriesSchema(workflow?: workflows.Workflow): SchemaStatus {
  const s3 = AWS.S3.use()
  const schemaUrl = workflow?.entriesSchema || ''
  const req = React.useCallback(() => objectSchema({ s3, schemaUrl }), [schemaUrl, s3])
  const result = Request.use(req, !!schemaUrl)

  if (!schemaUrl) return Ready()

  if (result === Request.Idle) return Idle
  if (result === Request.Loading) return Loading
  if (result instanceof Error) return Err(result)

  return Ready(result)
}
