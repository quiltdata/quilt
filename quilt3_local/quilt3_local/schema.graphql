directive @admin on FIELD_DEFINITION

scalar Datetime

scalar Json

scalar JsonRecord

type ContentIndexingSettings {
  extensions: [String!]!
  bytesDefault: Int!
  bytesMin: Int!
  bytesMax: Int!
}

type Config {
  contentIndexingSettings: ContentIndexingSettings! @admin
}

type BucketConfig {
  name: String!
  title: String!
  iconUrl: String
  description: String
  linkedData: Json
  overviewUrl: String
  tags: [String!]
  relevanceScore: Int!
  lastIndexed: Datetime
  snsNotificationArn: String @admin
  scannerParallelShardsDepth: Int @admin
  skipMetaDataIndexing: Boolean @admin
  fileExtensionsToIndex: [String!] @admin
  indexContentBytes: Int @admin
  permissions: [RoleBucketPermission!]! @admin
}

type UnmanagedRole {
  id: ID!
  name: String!
  arn: String
}

type ManagedRole {
  id: ID!
  name: String!
  arn: String
  permissions: [RoleBucketPermission!]!
}

union Role = UnmanagedRole | ManagedRole

enum BucketPermissionLevel {
  READ
  READ_WRITE
}

type RoleBucketPermission {
  role: Role!
  bucket: BucketConfig!
  level: BucketPermissionLevel
}

type AccessCountForDate {
  date: Datetime!
  value: Int!
}

type AccessCounts {
  total: Int!
  counts: [AccessCountForDate!]!
}

type PackageDir {
  path: String!
  metadata: JsonRecord
  # 32-bit Int is too small to represent dir size which quite often exceeds several GB
  size: Float!
  children: [PackageEntry!]!
}

type PackageFile {
  path: String!
  metadata: JsonRecord
  # 32-bit Int is too small to represent file size which quite often exceeds several GB
  size: Float!
  physicalKey: String!
}

union PackageEntry = PackageFile | PackageDir

type PackageRevision {
  hash: String!
  modified: Datetime!
  message: String
  metadata: JsonRecord!
  userMeta: JsonRecord
  totalEntries: Int
  # 32-bit Int is too small to represent package size which quite often exceeds several GB
  totalBytes: Float
  dir(path: String!): PackageDir
  file(path: String!): PackageFile
  accessCounts(window: Int = 30): AccessCounts
}

type PackageRevisionList {
  total: Int!
  page(number: Int = 1, perPage: Int = 10): [PackageRevision!]!
}

type Package {
  bucket: String!
  name: String!
  modified: Datetime!
  revisions: PackageRevisionList!
  revision(hashOrTag: String = "latest"): PackageRevision
  accessCounts(window: Int = 30): AccessCounts
}

enum PackageListOrder {
  NAME
  MODIFIED
}

type PackageList {
  total: Int!
  page(
    number: Int = 1
    perPage: Int = 30
    order: PackageListOrder = NAME
  ): [Package!]!
}

type Query {
  config: Config!
  bucketConfigs: [BucketConfig!]!
  bucketConfig(name: String!): BucketConfig
  packages(bucket: String!, filter: String): PackageList
  package(bucket: String!, name: String!): Package
  roles: [Role!]! @admin
  role(id: ID!): Role @admin
}

input BucketAddInput {
  name: String!
  title: String!
  iconUrl: String
  description: String
  linkedData: Json
  overviewUrl: String
  tags: [String!]
  relevanceScore: Int
  snsNotificationArn: String
  scannerParallelShardsDepth: Int
  skipMetaDataIndexing: Boolean
  fileExtensionsToIndex: [String!]
  indexContentBytes: Int
  setVersioning: Boolean
  delayScan: Boolean
}

input BucketUpdateInput {
  title: String!
  iconUrl: String
  description: String
  linkedData: Json
  overviewUrl: String
  tags: [String!]
  relevanceScore: Int
  snsNotificationArn: String
  scannerParallelShardsDepth: Int
  skipMetaDataIndexing: Boolean
  fileExtensionsToIndex: [String!]
  indexContentBytes: Int
  setVersioning: Boolean
}

type BucketAddSuccess {
  bucketConfig: BucketConfig!
}

type BucketNotFound {
  _: Boolean
}

type BucketAlreadyAdded {
  _: Boolean
}

type BucketDoesNotExist {
  _: Boolean
}

type BucketFileExtensionsToIndexInvalid {
  _: Boolean
}

type BucketIndexContentBytesInvalid {
  _: Boolean
}

type SnsInvalid {
  _: Boolean
}

type NotificationConfigurationError {
  _: Boolean
}

type NotificationTopicNotFound {
  _: Boolean
}

type InsufficientPermissions {
  _: Boolean
}

type IndexingInProgress {
  _: Boolean
}

union BucketAddResult =
    BucketAddSuccess
  | BucketAlreadyAdded
  | BucketDoesNotExist
  | BucketFileExtensionsToIndexInvalid
  | BucketIndexContentBytesInvalid
  | InsufficientPermissions
  | NotificationConfigurationError
  | NotificationTopicNotFound
  | SnsInvalid

type BucketUpdateSuccess {
  bucketConfig: BucketConfig!
}

union BucketUpdateResult =
    BucketUpdateSuccess
  | BucketFileExtensionsToIndexInvalid
  | BucketIndexContentBytesInvalid
  | BucketNotFound
  | NotificationConfigurationError
  | NotificationTopicNotFound
  | SnsInvalid

type BucketRemoveSuccess {
  _: Boolean
}

union BucketRemoveResult =
    BucketRemoveSuccess
  | BucketNotFound
  | IndexingInProgress

input PermissionInput {
  bucket: String!
  level: BucketPermissionLevel
}

input UnmanagedRoleInput {
  name: String!
  arn: String!
}

input ManagedRoleInput {
  name: String!
  permissions: [PermissionInput!]!
}

type RoleNameReserved {
  _: Boolean
}

type RoleNameExists {
  _: Boolean
}

type RoleNameInvalid {
  _: Boolean
}

type BucketConfigDoesNotExist {
  name: String!
}

type RoleDoesNotExist {
  _: Boolean
}

type RoleIsManaged {
  _: Boolean
}

type RoleIsUnmanaged {
  _: Boolean
}

type RoleCreateSuccess {
  role: Role!
}

union RoleCreateResult =
    RoleCreateSuccess
  | RoleNameReserved
  | RoleNameExists
  | RoleNameInvalid
  | BucketConfigDoesNotExist

type RoleUpdateSuccess {
  role: Role!
}

union RoleUpdateResult =
    RoleUpdateSuccess
  | RoleNameReserved
  | RoleNameExists
  | RoleNameInvalid
  | RoleIsManaged
  | RoleIsUnmanaged
  | BucketConfigDoesNotExist

type RoleDeleteSuccess {
  _: Boolean
}

type RoleAssigned {
  _: Boolean
}

union RoleDeleteResult =
    RoleDeleteSuccess
  | RoleDoesNotExist
  | RoleNameReserved
  | RoleAssigned

type Mutation {
  bucketAdd(input: BucketAddInput!): BucketAddResult! @admin
  bucketUpdate(name: String!, input: BucketUpdateInput!): BucketUpdateResult!
    @admin
  bucketRemove(name: String!): BucketRemoveResult! @admin

  roleCreateManaged(input: ManagedRoleInput!): RoleCreateResult! @admin
  roleCreateUnmanaged(input: UnmanagedRoleInput!): RoleCreateResult! @admin
  roleUpdateManaged(id: ID!, input: ManagedRoleInput!): RoleUpdateResult! @admin
  roleUpdateUnmanaged(id: ID!, input: UnmanagedRoleInput!): RoleUpdateResult!
    @admin
  roleDelete(id: ID!): RoleDeleteResult! @admin
}
