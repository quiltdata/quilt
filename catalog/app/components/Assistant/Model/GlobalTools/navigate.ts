import * as Eff from 'effect'
import * as RR from 'react-router-dom'
import { Schema as S } from '@effect/schema'

import * as Content from '../Content'
import * as Tool from '../Tool'

import { NavigableRouteSchema, navigate } from '../navigation'

export const NavigateSchema = S.Struct({
  route: NavigableRouteSchema,
}).annotations({
  title: 'navigate the catalog',
  description: 'navigate to a provided route',
})

export function useNavigate() {
  const history = RR.useHistory()
  return Tool.useMakeTool(
    NavigateSchema,
    ({ route }) =>
      Eff.pipe(
        navigate(route, history),
        Eff.Effect.match({
          onSuccess: () =>
            Tool.succeed(Content.text(`Navigating to the '${route.name}' route.`)),
          onFailure: (e) =>
            Tool.fail(
              Content.text(`Failed to navigate to the '${route.name}' route: ${e}`),
            ),
        }),
        Eff.Effect.map(Eff.Option.some),
      ),
    [history],
  )
}
