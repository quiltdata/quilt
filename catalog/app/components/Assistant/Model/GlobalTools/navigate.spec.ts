import { JSONSchema } from '@effect/schema'

import * as Nav from './navigate'

jest.mock(
  'constants/config',
  jest.fn(() => ({})),
)

describe('components/Assistant/Model/GlobalTools/navigate', () => {
  describe('NavigateSchema', () => {
    describe('produced JSON Schema', () => {
      it('should match the snapshot', () => {
        const jsonSchema = JSONSchema.make(Nav.NavigateSchema)
        expect(jsonSchema).toMatchSnapshot()
      })
    })
  })
})
