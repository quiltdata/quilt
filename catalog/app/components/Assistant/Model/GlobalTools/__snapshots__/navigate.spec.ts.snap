// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Assistant/Model/GlobalTools/navigate NavigateSchema produced JSON Schema should match the snapshot 1`] = `
{
  "$defs": {
    "BooleanPredicateState": {
      "additionalProperties": false,
      "properties": {
        "false": {
          "type": "boolean",
        },
        "true": {
          "type": "boolean",
        },
      },
      "required": [],
      "title": "Boolean predicate state",
      "type": "object",
    },
    "BooleanTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "const": "Boolean",
        },
        "value": {
          "$ref": "#/$defs/BooleanPredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
    "DatetimePredicateState": {
      "additionalProperties": false,
      "properties": {
        "gte": {
          "description": "ISO date string",
          "format": "date-time",
          "type": "string",
        },
        "lte": {
          "description": "ISO date string",
          "format": "date-time",
          "type": "string",
        },
      },
      "required": [],
      "title": "Datetime predicate state",
      "type": "object",
    },
    "DatetimeTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "const": "Datetime",
        },
        "value": {
          "$ref": "#/$defs/DatetimePredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
    "KeywordEnumPredicateState": {
      "additionalProperties": false,
      "properties": {
        "terms": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
      },
      "required": [],
      "title": "KeywordEnum predicate state",
      "type": "object",
    },
    "KeywordEnumTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "const": "KeywordEnum",
        },
        "value": {
          "$ref": "#/$defs/KeywordEnumPredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
    "KeywordWildcardPredicateState": {
      "additionalProperties": false,
      "properties": {
        "wildcard": {
          "type": "string",
        },
      },
      "required": [],
      "title": "KeywordWildcard predicate state",
      "type": "object",
    },
    "KeywordWildcardTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "const": "KeywordWildcard",
        },
        "value": {
          "$ref": "#/$defs/KeywordWildcardPredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
    "NumberPredicateState": {
      "additionalProperties": false,
      "properties": {
        "gte": {
          "type": "number",
        },
        "lte": {
          "type": "number",
        },
      },
      "required": [],
      "title": "Number predicate state",
      "type": "object",
    },
    "NumberTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "const": "Number",
        },
        "value": {
          "$ref": "#/$defs/NumberPredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
    "TextPredicateState": {
      "additionalProperties": false,
      "properties": {
        "queryString": {
          "type": "string",
        },
      },
      "required": [],
      "title": "Text predicate state",
      "type": "object",
    },
    "TextTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "const": "Text",
        },
        "value": {
          "$ref": "#/$defs/TextPredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
  },
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "description": "navigate to a provided route",
  "properties": {
    "route": {
      "anyOf": [
        {
          "additionalProperties": false,
          "description": "Home page",
          "properties": {
            "name": {
              "const": "home",
            },
            "params": {
              "$id": "/schemas/{}",
              "oneOf": [
                {
                  "type": "object",
                },
                {
                  "type": "array",
                },
              ],
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "description": "Installation page",
          "properties": {
            "name": {
              "const": "install",
            },
            "params": {
              "$id": "/schemas/{}",
              "oneOf": [
                {
                  "type": "object",
                },
                {
                  "type": "array",
                },
              ],
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "description": "Search page",
          "properties": {
            "name": {
              "const": "search",
            },
            "params": {
              "additionalProperties": false,
              "properties": {
                "buckets": {
                  "description": "a list of buckets to search in (keep empty to search in all buckets)",
                  "items": {
                    "type": "string",
                  },
                  "title": "search buckets",
                  "type": "array",
                },
                "order": {
                  "$comment": "/schemas/enums",
                  "description": "order of search results",
                  "oneOf": [
                    {
                      "const": "BEST_MATCH",
                      "title": "BEST_MATCH",
                    },
                    {
                      "const": "NEWEST",
                      "title": "NEWEST",
                    },
                    {
                      "const": "OLDEST",
                      "title": "OLDEST",
                    },
                  ],
                  "title": "search order",
                },
                "params": {
                  "anyOf": [
                    {
                      "additionalProperties": false,
                      "properties": {
                        "filter": {
                          "items": {
                            "anyOf": [
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "modified",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/DatetimePredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "last modified date",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "size",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/NumberPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "file size in bytes",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "name",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/KeywordWildcardPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "package name (aka namespace)",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "hash",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/KeywordWildcardPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "package revision hash",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "entries",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/NumberPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "number of package entries",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "comment",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/TextPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "package revision comment (aka commit message)",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "workflow",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/KeywordEnumPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "package workflow",
                                "type": "object",
                              },
                            ],
                          },
                          "title": "result filters (system)",
                          "type": "array",
                        },
                        "resultType": {
                          "const": "p",
                          "title": "result type: Quilt Package",
                        },
                        "userMetaFilters": {
                          "description": "filter results by user metadata",
                          "items": {
                            "additionalProperties": false,
                            "properties": {
                              "path": {
                                "type": "string",
                              },
                              "predicate": {
                                "anyOf": [
                                  {
                                    "$ref": "#/$defs/BooleanTaggedPredicate",
                                  },
                                  {
                                    "$ref": "#/$defs/DatetimeTaggedPredicate",
                                  },
                                  {
                                    "$ref": "#/$defs/KeywordEnumTaggedPredicate",
                                  },
                                  {
                                    "$ref": "#/$defs/KeywordWildcardTaggedPredicate",
                                  },
                                  {
                                    "$ref": "#/$defs/NumberTaggedPredicate",
                                  },
                                  {
                                    "$ref": "#/$defs/TextTaggedPredicate",
                                  },
                                ],
                              },
                            },
                            "required": [
                              "path",
                              "predicate",
                            ],
                            "type": "object",
                          },
                          "title": "result filters (user metadata)",
                          "type": "array",
                        },
                      },
                      "required": [
                        "resultType",
                      ],
                      "title": "package-specific search parameters",
                      "type": "object",
                    },
                    {
                      "additionalProperties": false,
                      "properties": {
                        "filter": {
                          "items": {
                            "anyOf": [
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "modified",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/DatetimePredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "last modified date",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "size",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/NumberPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "file size in bytes",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "ext",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/KeywordEnumPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "file extensions (with a leading dot)",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "key",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/KeywordWildcardPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "file name (aka S3 Object Key)",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "content",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/TextPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "indexed text contents",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "const": "deleted",
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/BooleanPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "whether a file is a delete marker",
                                "type": "object",
                              },
                            ],
                          },
                          "title": "result filters",
                          "type": "array",
                        },
                        "resultType": {
                          "const": "o",
                          "title": "result type: S3 Object",
                        },
                      },
                      "required": [
                        "resultType",
                      ],
                      "title": "object-specific search parameters",
                      "type": "object",
                    },
                  ],
                  "title": "result type-specific parameters",
                },
                "searchString": {
                  "description": "string to search for, ElasticSearch syntax supported",
                  "title": "search string",
                  "type": "string",
                },
              },
              "required": [
                "params",
              ],
              "type": "object",
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "description": "TBD",
          "properties": {
            "name": {
              "const": "activate",
            },
            "params": {
              "$id": "/schemas/{}",
              "oneOf": [
                {
                  "type": "object",
                },
                {
                  "type": "array",
                },
              ],
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "description": "Bucket root page",
          "properties": {
            "name": {
              "const": "bucket",
            },
            "params": {
              "$id": "/schemas/{}",
              "oneOf": [
                {
                  "type": "object",
                },
                {
                  "type": "array",
                },
              ],
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
      ],
    },
  },
  "required": [
    "route",
  ],
  "title": "navigate the catalog",
  "type": "object",
}
`;
